gpytorch/lazy/lazy_tensor.py

```diff
@@ -732,6 +734,217 @@ def add_jitter(self, jitter_val=1e-3):
         diag = torch.tensor(jitter_val, dtype=self.dtype, device=self.device)
         return self.add_diag(diag)
 
+    def cat_rows(self, cross_mat, new_mat, generate_roots=True):
+        """
+        Concatenates new rows and columns to the matrix that this LazyTensor represents, e.g.
+        C = ((A; B); B^T, D). where A is the existing lazy tensor, and B (cross_mat) and D (new_mat)
+        are new components. This is most commonly used when fantasizing with kernel matrices.
+
+        We have access to K \\approx LL^T and K^{-1} \\approx RR^T, where L and R are low rank matrices
+        resulting from root and root inverse decompositions (see the LOVE paper).
+
+        To update R, we first update L:
+            [K U; U^T S] = [L 0; A B][L^T A^T; 0 BT]
+        Solving this matrix equation, we get:
+            K = LL^T ==>       L = L
+            U = LA^T ==>       A = UR^{-1}
+            S = AA^T + BB^T ==> B = cholesky(S - AA^T)
+
+        Once we've computed Z = [L 0; A B], we have that the new kernel matrix [K U; U^T S] \approx ZZ^T. Therefore,
+        we can form a pseudo-inverse of Z directly to approximate [K U; U^T S]^{-1/2}.
+
+        This strategy is also described in "Efficient Nonmyopic Bayesian Optimization via One-Shot Multistep Trees,"
+        Jiang et al, NeurIPS, 2020. https://arxiv.org/abs/2006.15779.
+
+        Args:
+            cross_mat (:obj:`torch.tensor`): the matrix :math:`B` we are appending to the matrix :math:`A`.
+                If :math:`A` is n x n, then this matrix should be n x k.
+            new_mat (:obj:`torch.tensor`): the matrix :math:`D` we are appending to the matrix :math:`A`.
+                If :math:`B` is n x k, then this matrix should be k x k.
+            generate_roots (:obj:`bool`): whether to generate the root decomposition of :math:`A` even if it
+                has not been created yet.
+
+        Returns:
+            :obj:`LazyTensor`: concatenated lazy tensor with the new rows and columns.
+        """
+        from . import lazify
+        from .batch_repeat_lazy_tensor import BatchRepeatLazyTensor
+        from .cat_lazy_tensor import CatLazyTensor
+        from .root_lazy_tensor import RootLazyTensor
+        from .triangular_lazy_tensor import TriangularLazyTensor
+
+        cross_mat = lazify(cross_mat)
+        batch_shape = cross_mat.shape[:-2]
```

Comment:
Is it worth it to make sure that self, cross_mat, and new_mat all have the same batch shape (using the broadcasting helper)?

Reply:
I don't think so. In general, we want to be able to have self be non-batched but cross_mat and new_mat to be batched. That is something like this currently works
```python
x = torch.randn(10, 5)
a = lazify(x.t() @ x) # 5 x 5 full rank

newx = torch.randn(32, 7, 5)
newa = lazify(newx.transpose(-1, -2) @ newx)

crossa = 0.01 * torch.randn(32, 5, 5)
result = a.cat_rows(crossa, newa) # is of shape 32 x 5 x 5
```

Reply:
And a non-batched cross matrix fails currently, e.g. 
```python
cross = 0.01 * torch.randn(5, 5)
result = a.cat_rows(crossa, newa) # failed
```
which is also the desired behavior.

Reply:
Just saw this resolved one. Didn't age all that well given my comment here: https://github.com/cornellius-gp/gpytorch/pull/1499/files#r594720889

let me see if I can put together a fix for that.